### ISSUE TITLE
Interesting behavior with metadata queries

### ISSUE THREAD

**Issue by korydraughn** [irods/irods_api_plugin_genquery2#42]
## Interesting Behavior

Notice the results produced by the following snippet.
```bash
itouch foo
imeta add -d foo alice jones
imeta add -C . bob jones
imeta add -C . job developer
iquery "select META_COLL_ATTR_NAME" | jq
[
  [
    "bob"
  ],
  [
    "job"
  ],
  [
    ""    <------ Look suspicious?
  ]
]
```

Here's the SQL that produced this result.
```sql
SELECT DISTINCT
    mmc.meta_attr_name
FROM
    R_COLL_MAIN t0
    INNER JOIN R_OBJT_ACCESS pcoa ON t0.coll_id = pcoa.object_id
    INNER JOIN R_TOKN_MAIN pct ON pcoa.access_type_id = pct.token_id
    INNER JOIN R_USER_MAIN pcu ON pcoa.user_id = pcu.user_id
    LEFT JOIN R_OBJT_METAMAP ommc ON t0.coll_id = ommc.object_id
    LEFT JOIN R_META_MAIN mmc ON ommc.meta_id = mmc.meta_id
WHERE
    pcoa.access_type_id >= 1000 FETCH FIRST 16 ROWS ONLY
```

Turns out the results are correct due to the left joins. This query will produce empty strings for ALL collections that do NOT have metadata attached to them.

## Solution

To exclude these rows, you must filter the results using a condition like so.
```bash
iquery "select META_COLL_ATTR_NAME where META_COLL_ATTR_NAME is not null" | jq
[
  [
    "bob"
  ],
  [
    "job"
  ]
]
```

## Thoughts

This was surprising at first glance, but the behavior is explained due to the SQL.

This does not limit a user's ability to use GenQuery2, it simply means users have to get use to the behavior and learn how to use it appropriately. Also, documentation will help with understanding.

Overall, the behavior is acceptable IMO.

**Comment by trel** [comment]
yes, this feels like a documentation exercise.    with the increased expressiveness of genquery2, this feels acceptable to me too.

think of a section name to describe this 'new' behavior...

**Comment by korydraughn** [comment]
We can also add options that instruct the parser to use inner joins. Inner joins remove the need to add a filtering condition, but at the cost of only allowing users to query metadata for one type of entity in iRODS.

**Comment by trel** [comment]
One of the greatest limitations of GenQuery1 is the lack of ability to return rows that don't match ALL the things requested (and lack of OR support... which is related...).    This documentation and filtering capabilities open up more possibilities, but users will need to have somewhere to go to see lots of examples.

A home for lots of examples would be nice.   Maybe that's just docs.irods.org in the end...

**Comment by korydraughn** [comment]
Sure. A page/section in the docs sounds like the right approach.

**Comment by trel** [comment]
Closing this issue.  This repository is being archived.

Since iRODS 4.3.2, GenQuery2 has been available in the server.
